### DISCUSSION TITLE
Recording videos using "PlusServerRemoteControl.exe"

### DISCUSSION THREAD

**Discussion by lbo2021** [PlusToolkit/PlusLib#892] [Q&A]
Hi,

I'm trying to record videos using PlusServerRemoteControl.exe but I'm getting those error messages and the recorded videos are lagging!

Also if I reduce the sampling rate I get the same error.

If I record a single video, if the resolution is maximum 1920x1080 it works well also with a sampling rate of 30fps. If I record a video with an higher resolution it lags even with a single video

|WARNING|3291.490000|SERVER> Recording of frames takes too long time (1.78sec instead of the allocated 0.0333333sec, recording lags by 1.80967sec). This can cause slow-down of the application and non-uniform sampling. Reduce the acquisition rate or sampling rate to resolve the problem.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(405)
|WARNING|3296.126000|SERVER> Recording of frames takes too long time (4.57sec instead of the allocated 0.0333333sec, recording lags by 4.74667sec). This can cause slow-down of the application and non-uniform sampling. Reduce the acquisition rate or sampling rate to resolve the problem.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(405)
|ERROR|3296.144000|SERVER> Recording cannot keep up with the acquisition. Skip 4.74667 seconds of the data stream to catch up.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(420)
|WARNING|3298.512000|SERVER> Recording of frames takes too long time (1.556sec instead of the allocated 0.0333333sec, recording lags by 1.55067sec). This can cause slow-down of the application and non-uniform sampling. Reduce the acquisition rate or sampling rate to resolve the problem.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(405)
|WARNING|3302.713000|SERVER> Recording of frames takes too long time (4.165sec instead of the allocated 0.0333333sec, recording lags by 4.15067sec). This can cause slow-down of the application and non-uniform sampling. Reduce the acquisition rate or sampling rate to resolve the problem.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(405)
|ERROR|3302.715000|SERVER> Recording cannot keep up with the acquisition. Skip 4.15067 seconds of the data stream to catch up.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(420)
|WARNING|3316.612000|SERVER> Recording of frames takes too long time (1.01sec instead of the allocated 0.0333333sec, recording lags by 1.00667sec). This can cause slow-down of the application and non-uniform sampling. Reduce the acquisition rate or sampling rate to resolve the problem.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(405)
|WARNING|3316.613000|SERVER> Recording of frames takes too long time (2.728sec instead of the allocated 0.0333333sec, recording lags by 2.735sec). This can cause slow-down of the application and non-uniform sampling. Reduce the acquisition rate or sampling rate to resolve the problem.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(405)
|WARNING|3316.614000|SERVER> Recording of frames takes too long time (4.178sec instead of the allocated 0.0333333sec, recording lags by 5.38sec). This can cause slow-down of the application and non-uniform sampling. Reduce the acquisition rate or sampling rate to resolve the problem.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(405)
|ERROR|3316.614000|SERVER> Recording cannot keep up with the acquisition. Skip 5.38 seconds of the data stream to catch up.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(420)
|WARNING|3316.615000|SERVER> Recording of frames takes too long time (1.447sec instead of the allocated 0.0333333sec, recording lags by 1.427sec). This can cause slow-down of the application and non-uniform sampling. Reduce the acquisition rate or sampling rate to resolve the problem.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(405)
|WARNING|3317.164000|SERVER> Recording of frames takes too long time (3.375sec instead of the allocated 0.0333333sec, recording lags by 3.63567sec). This can cause slow-down of the application and non-uniform sampling. Reduce the acquisition rate or sampling rate to resolve the problem.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(405)
|ERROR|3317.165000|SERVER> Recording cannot keep up with the acquisition. Skip 3.63567 seconds of the data stream to catch up.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(420)
|WARNING|3319.116000|SERVER> Recording of frames takes too long time (1.221sec instead of the allocated 0.0333333sec, recording lags by 1.251sec). This can cause slow-down of the application and non-uniform sampling. Reduce the acquisition rate or sampling rate to resolve the problem.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(405)

I'm running those commands to start and stop the recording:

PlusServerRemoteControl.exe --command=START_ACQUISITION --device="CaptureVideo1" --output-file="RecordedVideo1.mkv"
PlusServerRemoteControl.exe --command=START_ACQUISITION --device="CaptureVideo2" --output-file="RecordedVideo2.mkv"

PlusServerRemoteControl.exe --command=STOP_ACQUISITION --device="CaptureVideo1"
PlusServerRemoteControl.exe --command=STOP_ACQUISITION --device="CaptureVideo2"

Do you have any idea how to solve this issue?

**Comment by adamrankin** [comment]
Are you recording to an SSD? Are you using compression?

**Reply by lbo2021** [reply]
Recording to SSD without compression


If I try to enable compression, I get this warning:

|WARNING|7778.618000|SERVER> Compressed saving of metaimage file requested. This is not supported. Reverting to uncompressed metaimage file.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(191)
|WARNING|7778.660000|SERVER> Compressed saving of metaimage file requested. This is not supported. Reverting to uncompressed metaimage file.| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\VirtualDevices\vtkPlusVirtualCapture.cxx(191)

**Reply by adamrankin** [reply]
Oh, yes, higher than 1080. I don't think it will be possible in real time. You can try compression, but I don't know if the CPU will be able to keep up. Are you trying to record 4k?

**Reply by adamrankin** [reply]
Yes, try recording to nrrd format, this allows for compression in real-time.

**Reply by lbo2021** [reply]
I just tried to record 4K to make a test but that is not my main goal.

My goal is to be able to record 2 videos 1920x1080 and some tracking data at the same time.
I need all these recording synchronized. 
I would like to record each video in .mkv format and the tracking data into a .igs.mha file.

I tried to record to nrrd format and I am not getting the error anymore but the output file is really huge.. some seconds of a 
 1920x1080 recording is bigger than 1 GB also with compression

Any suggestion how to record the data I described keeping the output files in a reasonable size?

**Reply by adamrankin** [reply]
How long do you need to record?

**Reply by lbo2021** [reply]
Let's say 1 hour.  It would generate a .nrrd file with a size of around 120 GB (for each video)!

**Reply by adamrankin** [reply]
Is that not an option? You could post-process into another format afterwards then delete the originals.

**Reply by adamrankin** [reply]
Note, Plus doesn't do recording encoding, just compression (real-time zip basically). You won't get web format videos as output from Plus.

**Reply by lbo2021** [reply]
Yes you are right, it is an option.

However, @adamrankin  I noticed strange behaviors if I record for some minutes. After some minutes the recording is corrupted. The issue I encountered are:

- Wrong colors after some minutes
-  Image transformed

Please see at this link a video that show the issues I encountered https://drive.google.com/file/d/1lNBy-w3FqKyTLfSMq6tnbMu8NuuLVOU1/view?usp=sharing

This video is a replay on Slicer of a video recorded using "PlusServerRemoteControl.exe". Please look at the last seconds.
The issue is generated during the recording process. I'm sure that the input video is good because while I was recording I was at the same time connected to the Plus Server with Slicer, so I could check the video and it did not have any error
